### ISSUE TITLE
GetObject api with default http client hangs while downloading large file

### ISSUE THREAD

**Issue by sdhuka** [awslabs/aws-sdk-kotlin#525]
### Describe the bug

When downloading large files(~120mb), the GetObject API with the default http client hangs, it downloads few mb and then doesn't rest of the file.
PS: Same api works as expected when used with Ktor client.

### Expected behavior

GetObject api should download the entire object without any pause.

### Current behavior

GetObject api with default http client pauses download after downloading few bytes while downloading large files. It is able to download small files correctly.

### Steps to Reproduce

1) Upload large file(>100mb) to S3 using AWS Console
2) Download file using GetObject api

        s3.getObject(GetObjectRequest {
                        key = INVALID_KEY
                        bucket = BUCKEY_NAME
                    }) { response -> ... 
        
         }
3) Observe the file progress using the device file manager on AS

### AWS Kotlin SDK version used

0.12.0-beta

### Platform (JVM/JS/Native)

Native

### Operating System and version

Android 31

**Comment by kggilmer** [comment]
I ran into a somewhat different issue when attempting to reproduce this.  I adapted an existing test to write a 120Mb file to an S3 bucket.  I then used the following code to attempt to download the file.  The program ran for several minutes but eventually completed with:

```
aws.sdk.kotlin.runtime.ClientException: CrtHttpEngine::response failed: ec=1051; description=socket is closed.
	at app//aws.sdk.kotlin.runtime.http.engine.crt.SdkStreamResponseHandler.onResponseComplete(SdkStreamResponseHandler.kt:151)
	at app//aws.sdk.kotlin.crt.http.HttpRequestUtilKt$asJniStreamResponseHandler$1.onResponseComplete(HttpRequestUtil.kt:59)
	at app//software.amazon.awssdk.crt.http.HttpStreamResponseHandlerNativeAdapter.onResponseComplete(HttpStreamResponseHandlerNativeAdapter.java:33)
```

The file was partially written to disk:

```bash
$ ls -hl /tmp/big-file-s3.txt 
-rw-r--r-- 1 kggilmer domain^users 8.2M Feb 10 11:48 /tmp/big-file-s3.txt
```

I confirmed that the uploaded file was 120Mb in size.  Test program:

```kotlin
    fun testPutObjectFromFile(): Unit = runBlocking {
        val keyName = "big-file.txt"

        val req = GetObjectRequest {
            bucket = testBucket
            key = keyName
        }
        client.getObject(req) {
            it.body?.writeToFile(File("/tmp/big-file-s3.txt"))
        }
    }
```

I then looked to verify if the `Ktor` client does work.  I used this test:

```kotlin
    @Test
    fun testReadBigObjectFromFile(): Unit = runBlocking {
        val config = AwsClientConfig.fromEnvironment()

        S3Client(config) {
            httpClientEngine = KtorEngine()
        }.use { client ->
            client.getObject(GetObjectRequest {
                key = "big-file.txt"
                bucket = "my-test-bucket"
            }) {
                it.body?.writeToFile(File("/tmp/big-file-s3.txt"))
            }
        }
    }
```

And I consistently received this exception:

```
Max send count 20 exceeded. Consider increasing the property maxSendCount if more is required.
io.ktor.client.features.SendCountExceedException: Max send count 20 exceeded. Consider increasing the property maxSendCount if more is required.
	at app//io.ktor.client.features.HttpSend$DefaultSender.execute(HttpSend.kt:121)
	at app//io.ktor.client.features.HttpRedirect$Feature.handleCall(HttpRedirect.kt:95)
	at app//io.ktor.client.features.HttpRedirect$Feature.access$handleCall(HttpRedirect.kt:47)
	at app//io.ktor.client.features.HttpRedirect$Feature$handleCall$1.invokeSuspend(HttpRedirect.kt)
...
Caused by: io.ktor.client.features.SendCountExceedException: Max send count 20 exceeded. Consider increasing the property maxSendCount if more is required.
	at app//io.ktor.client.features.HttpSend$DefaultSender.execute(HttpSend.kt:121)
	at app//io.ktor.client.features.HttpRedirect$Feature.handleCall(HttpRedirect.kt:95)
	at app//io.ktor.client.features.HttpRedirect$Feature.access$handleCall(HttpRedirect.kt:47)
```


In summary I was able to upload large files successfully to S3 however neither the CRT (default) HTTP client nor the Ktor client was able to download the file in my tests.

**Comment by ianbotsf** [comment]
It appears the `SendCountExceedException` is caused by attempting to read from an existing bucket without adequate permissions (e.g., a bucket owned by another AWS account). Spinning off as separate issue #527.

**Comment by ianbotsf** [comment]
The fix for this has been approved and pushed to **main**. It should be available in the next SDK release.
